bochs/iodev/usb/usb_uasp.cc

```diff
@@ -0,0 +1,512 @@
+	Bit16u tag;        // big endian
+	Bit8u  prio_attr;  // Task Attribute
+	Bit8u  rsvd1;      //
+	Bit8u  len;        // len (if more than 16 bytes) (i.e.: if command len is 20 bytes, this field is 4)  (bottom 2 bits are reserved)
+	Bit8u  rsvd2;      //
+	Bit64u lun;        // eight byte lun
+  Bit8u  com_block[16]; // command block
+};
+
+struct S_UASP_STATUS {
+	Bit8u  id;         // usually IU_SENSE ?
+	Bit8u  rsvd0;
+	Bit16u tag;        // big endian
+	Bit16u stat_qual;  // big endian
+	Bit8u  status;
+	Bit8u  resv1[7];
+	Bit16u len;        // big endian
+  Bit8u  sense[18];
+};
+
+#define UASP_FROM_COMMAND  ((Bit32u) -1)
+#define U_NONE          0
+#define U_SRV_ACT   (1<<0)
+#define U_IS_LBA    (1<<1)
+struct S_UASP_INPUT {
+  Bit8u  command;      // command byte
+  Bit8u  serv_action;  // command/service action
+  Bit8u  cmd_len;      // length of the command buf
+  Bit8u  direction;    // in or out
+  Bit8u  flags;        // see S_UASP_FLAGS_*
+  Bit32u data_len;     // if the command is fixed at a count of bytes, this holds that count (next two members are ignored)
+  int    offset;       // byte offset of data length requested
+  int    size;         // size in bytes of this field
+};
+
+int usb_msd_device_c::uasp_handle_data(USBPacket *p)
+{
+  int ret = 0;
+  Bit8u *data = p->data;
+  int len = p->len;
+  int index = p->strm_pid; // high-speed device will be zero. super-speed device will have the stream id.
+  
+  BX_INFO(("uasp_handle_data(): %X  ep=%d  index=%d  len=%d", p->pid, p->devep, index, len));
+  
+  switch (p->pid) {
+    case USB_TOKEN_OUT:
+      if (p->devep == MSD_UASP_DATAOUT) { // Bulk out pipe
+        if ((index < 0) || (index > UASP_MAX_STREAMS_N))
+          goto fail;
+        ret = uasp_process_request(p, index);
+        if (ret == USB_RET_ASYNC)
+          usb_defer_packet(p, this);
+      } else if (p->devep == MSD_UASP_COMMAND) { // Command pipe
+        ret = uasp_do_command(p);
+      }
+      break;
+      
+    case USB_TOKEN_IN:
+      // TODO: these are the same
+      if ((p->devep == MSD_UASP_DATAIN) || // Bulk in pipe
+          (p->devep == MSD_UASP_STATUS)) { // Status pipe
+        if ((index < 0) || (index > UASP_MAX_STREAMS_N))
+          goto fail;
+        ret = uasp_process_request(p, index);
+        if (ret == USB_RET_ASYNC)
+          usb_defer_packet(p, this);
+      }
+      break;
+      
+    default:
+      BX_ERROR(("USB MSD (UASP) handle_data: bad token"));
+fail:
+      d.stall = 1;
+      ret = USB_RET_STALL;
+  }
+
+  return ret;
+}
+
+void usb_msd_device_c::uasp_initialize_request(int index)
+{
+  UASPRequest *req = &s.uasp_request[index];
+
+  req->mode = UASP_SET_ACTIVE(1);  // setting to 1 clears other flags
+  req->data_len = 0;
+  req->result = 0;
+  req->tag = 0;
+  req->scsi_len = 0;
+  req->status = NULL;  
+  req->p = NULL;
+
+  d.stall = 0;
+}
+
+UASPRequest *usb_msd_device_c::uasp_find_request(Bit32u tag)
+{
+  int i;
+
+  for (i=0; i<UASP_MAX_STREAMS_N; i++) {
```

Comment:
int i could be inserted inside for()

Reply:
yes it can. Some other files do not, so I was just being consistent with them.

Reply:
Even if some other files badly written, it is better to avoid writing bad code :)

One of the obviously bad things is to define variable separately from its initialization in C++.
if possible it is always better to initialize variable in place of definition.
Loop counters better to define and initialize inside the loop